Sources/TokamakCore/Fiber/Passes/ReconcilePass.swift

```diff
@@ -223,41 +231,73 @@ struct ReconcilePass: FiberReconcilerPass {
-                let previous = node.fiber?.alternate?.element
-      {
-        if let fiber = node.fiber {
-          invalidateCache(for: fiber, in: reconciler, caches: caches)
-        }
-        // This is a completely different type of view.
-        return .replace(parent: parent, previous: previous, replacement: element)
-      } else if let newContent = node.newContent,
-                newContent != element.content
-      {
-        if let fiber = node.fiber {
-          invalidateCache(for: fiber, in: reconciler, caches: caches)
-        }
-        // This is the same type of view, but its backing data has changed.
-        return .update(
-          previous: element,
-          newContent: newContent,
-          geometry: node.fiber?.geometry ?? .init(
-            origin: .init(origin: .zero),
-            dimensions: .init(size: .zero, alignmentGuides: [:]),
-            proposal: .unspecified
-          )
+    guard let fiber = node.fiber else { return nil }
+
+    func canUpdate(_ fiber: FiberReconciler<R>.Fiber) -> Bool {
+      fiber.typeInfo?.type == fiber.alternate?.typeInfo?.type
+    }
+
+    invalidateCache(for: fiber, in: reconciler, caches: caches)
+
+    switch (fiber.element, node.newContent, fiber.alternate?.element) {
+    case let (nil, content?, nil):
+      guard let index = fiber.elementIndex, let parent = fiber.elementParent?.element else { break }
+
+      let el = R.ElementType(from: content)
+      fiber.element = el
+      fiber.alternate?.element = el
+
+      return .insert(element: el, parent: parent, index: index)
+
+    case let (nil, content?, altElement?) where !canUpdate(fiber):
+      guard let parent = fiber.elementParent?.element else { break }
+
+      let el = R.ElementType(from: content)
+      fiber.element = el
+      fiber.alternate?.element = el
+
+      return .replace(parent: parent, previous: altElement, replacement: el)
+
+    case let (nil, content?, element?) where canUpdate(fiber) && content != element.content,
+         let (element?, content?, _) where canUpdate(fiber) && content != element.content:
+      guard fiber.elementParent?.element != nil else { break } // todo: is this needed?
+
+      fiber.element = element
+      return .update(
+        previous: element,
+        newContent: content,
+        geometry: fiber.geometry ?? .init(
+          origin: .init(origin: .zero),
+          dimensions: .init(size: .zero, alignmentGuides: [:]),
+          proposal: .unspecified
         )
-      }
+      )
+
+    case let (_, nil, alt?):
+      guard let parent = fiber.alternate?.elementParent?.element else { break } // todo: name => altParent?
+
+      fiber.element = nil
+      fiber.elementIndex = 0
+      if let p = fiber.elementParent { caches.elementIndices[ObjectIdentifier(p)]? -= 1 }
+      return .remove(element: alt, parent: parent)
+
+    case let (element?, content, _) where fiber.alternate?.element == nil:
```

Comment:
Couldn't this be written as:
```suggestion
    case let (element?, content, nil):
```